### PR TITLE
Refactor Dotty Singleton Enum cases

### PR DESCRIPTION
**Pull request by bishabosha**
now we create module symbols for them which are erased as if it were typed an intersection of its parents, and accesses to the module value in byte code are instead an access of a field in the companion of the enum class

e.g.

```scala
trait Mixin
enum Colour {
  case Red extends Colour with Mixin
}
```

`Colour.Red` before erasure is the singleton type `Colour.Red.type` with its own module class.
After erasure `Colour.Red` will have the type of the erasure of `Colour with Mixin` and at bytecode is accessed by `getstatic Colour$.MODULE$.Red:LColour;`

### CODE REVIEW

File: src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala [hunk 1]
```diff
@@ -498,6 +498,9 @@ class TreeUnpickler[Tasty <: TastyUniverse](
           }
         }
       }.ensuring(isSymbol(_), s"${ctx.classRoot}: Could not create symbol at $start")
+      if (tag == VALDEF && flags.is(Enum)) {
+        sym.updateAttachment(new symbolTable.DottyEnumSingleton(sym.name.toString))
```
Comment (sjrd) [inline src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala]: I wonder if we could (intelligently) reuse the `JAVA_ENUM` flag for this. A actual *Java* enum would have `JAVA_ENUM` *and* `JAVA`, whereas a Dotty enum would only have `JAVA_ENUM`.

Perhaps you could audit uses of `JAVA_ENUM` / `isJavaEnum` to see whether that would make sense (simplify some code paths or not affect them) or not (making some code paths more complex).

That could definitely done in a separate PR, though.

**Review by sjrd — APPROVED** [review]

File: src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala (hunk 1 above)
Comment (bishabosha) [reply src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala]: if there was a way to get the original name of the symbol before it is mangled to an inner class name I'd be happy

Comment (sjrd) [reply src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala]: `originalName`?

Comment (bishabosha) [reply src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala]: no that is still an inner class name, e.g. `Colour$Red`

**Pull request merged**